File: 2. Structural Design Patterns/Bridge/Go/BridgeTemplate.go
package main

// Bridge Pattern – Go Template
// Separate Abstraction from Implementor so they can vary independently.

type Implementor interface{ OpImpl() }

type ConcreteImplA struct{}
func (ConcreteImplA) OpImpl() { /* ... */ }

type Abstraction struct{ impl Implementor }
func (a Abstraction) Operation(){ a.impl.OpImpl() }

type RefinedAbstraction struct{ Abstraction }

func main(){ a := RefinedAbstraction{Abstraction{impl: ConcreteImplA{}}}; a.Operation() }

File: 2. Structural Design Patterns/Bridge/Kotlin/BridgeTemplate.kt
// Bridge Pattern – Kotlin Template
// Separate Abstraction from Implementor so they can vary independently.

interface Implementor { fun opImpl() }
class ConcreteImplementorA : Implementor { override fun opImpl() { /* ... */ } }

open class Abstraction(protected val impl: Implementor) {
    open fun operation() { impl.opImpl() }
}

class RefinedAbstraction(impl: Implementor) : Abstraction(impl) {
    override fun operation() { super.operation() }
}

fun main(){ val a = RefinedAbstraction(ConcreteImplementorA()); a.operation() }
